File: House-Robber.cpp
/*
     int rob(vector<int>& nums) {
        int n = nums.size();
        vector<int> dp(n+1,0);
        dp[n] = 0 ;
        dp[n-1] = nums[n-1];
        for(int i = n-2 ;i>=0;i--){
            dp[i] = max(nums[i]+dp[i+2],dp[i+1]);
        }
        return dp[0];
    }
*/
class Solution {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        int prev2 = 0 ;
        int prev1 = nums[n-1];
        for(int i = n-2 ;i>=0;i--){
           int curr = max(nums[i]+prev2,prev1);
           prev2 = prev1;
           prev1 = curr;

        }
        return prev1;
    }
};

File: House-Robber.cs
/*
public class Solution {
    public int solveRecursive(int[] nums,int i){
        if(i>= nums.Length){
            return 0;
        }else{
            return Math.Max(nums[i]+solveRecursive(nums,i+2),solveRecursive(nums,i+1));
        }
    }
    public int Rob(int[] nums) {
        return Math.Max(solveRecursive(nums,0),solveRecursive(nums,1));
    }
}
*/
public class Solution {
    public int dpSolve(int n,int [] nums,int[] dp){
        if(n>= nums.Length){
            return 0 ;
        }
        if(dp[n] == -1){
            dp[n] = Math.Max(nums[n]+dpSolve(n+2,nums,dp),dpSolve(n+1,nums,dp));
        }
        return dp[n]; 
    }
    public int Rob(int[] nums) {
        int[] dp = new int[nums.Length];
        Array.Fill(dp,-1);
        return Math.Max(dpSolve(0,nums,dp),dpSolve(1,nums,dp));
    }
}